content/technotes/2020-05-28-rsvg2/index.md

```diff
@@ -0,0 +1,157 @@
+str(bitmap)
+##> raw [1:4, 1:600, 1:600] 
+
+# Read the bitmap in magick
+image <- magick::image_read(bitmap)
+```
+
+<svg viewBox="0 0 1200 250" xmlns="http://www.w3.org/2000/svg">
+  <style>
+    circle {
+      fill: gold;
+      stroke: maroon;
+      stroke-width: 12px;
+    }
+    text {
+      fill: navy;
+      font-size: 2em;
+      font-family: "Times, Serif"
+    }
+  </style>	
+  <circle cx="200" cy="125" r="120" />
+  <text x="140" y="40" transform="rotate(30 10,20)">I love SVG!</text>
+</svg>
+
+In magick, you can easily do all sorts of post-processing and conversion of the bitmap image. The `magick` package has a convenient wrapper function [read_image_svg](https://docs.ropensci.org/magick/reference/editing.html) that does exactly this: it uses `rsvg` to render the image and then reads the bitmap data as a magick image.
+
+## Using SVG for R graphics
+
+The best way to create svg files from graphics in R is using the [svglite](https://cran.r-project.org/web/packages/svglite/index.html) package. Try running the code below and then have a look at [mtcars.svg](mtcars.svg) in a text editor.
+
+
+```r
+library(svglite)
+library(ggplot2)
+
+# SVG sizes are in inches, not pixels
+res <- 144
+svglite("mtcars.svg", width = 1080/res, height = 720/res)
+ggplot(mtcars, aes(mpg, disp, colour = hp)) + geom_point() + geom_smooth()
+dev.off()
+```
+
+<object id="my-svg" type="image/svg+xml" data="mtcars.svg"></object>
+<script src="//cdn.jsdelivr.net/npm/vivus@latest/dist/vivus.min.js"></script>
+<script>
+var anim = new Vivus('my-svg', { duration: 200 });
+</script>
+
+Again we can use rsvg directly or via magick to convert this to a bitmap image:
+
+```r
+# Render the svg into a png image with rsvg via magick
+img <- magick::image_read_svg("mtcars.svg", width = 1080)
+magick::image_write(img, 'mtcars.png')
+```
+
+This generates a png image of with 1080x720px, without loss of quality.
+
+## Using CSS for R graphics?
+
+One feature in librsvg that has improved a lot from servo is support for CSS. As can be seen in the example above, svg allows for specifying global styling via CSS rules. In the browser, CSS and JavaScript can also be used to add interactivity and animation to SVG.
+
+With the latest version of librsvg it is now also possible to specify the CSS stylesheet from an external file, rather than inlining it in the svg itself. For example you can have a `fig.svg` file like this:
+
+```xml
+<svg viewBox="0 0 1200 250" xmlns="http://www.w3.org/2000/svg">
+  <circle cx="200" cy="125" r="120" />
+  <text x="140" y="40" transform="rotate(30 10,20)">Separate CSS!</text>
+</svg>
+```
+
+And a separate `style.css` file like this:
+
+```css
+circle {
+  fill: gold;
+  stroke: maroon;
+  stroke-width: 12px;
+}
+text {
+  fill: navy;
+  font-size: 2em;
+  font-family: "Times, Serif"
+}
+```
+
+Which you would render in R like this to get the same figure as above.
+
+```r
+rsvg_png('fig.svg', css = 'style.css', file = 'output.png')
```

Comment:
Not really relevant but it made me think of a recent convo about dark mode https://twitter.com/xieyihui/status/1263322909084471297 (you'd need to keep the svg on a web page)

Reply:
That's good point!